File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.js
/**
 * @param {number[][]} points
 * @return {number}
 */
var minCostConnectPoints = function(points) {
  //cretae adj list
  let N = points.length;
//   let adj = {};
  
//   for (let i = 0; i < N; i++) {
//     adj[i] = [];
//   }
  
//   for (let i = 0; i < N; i++) {
//     let [xi, yi] = points[i];
//     for (let j = i + 1; j < N; j++) {
//       let [xj, yj] = points[j];
//       let dist = Math.abs(xi - xj) + Math.abs(yi - yj);
//       adj[i].push([dist, j]);
//       adj[j].push([dist, i]);
//     }
//   }
  
  //minHeap + queue
  let res = 0;
  let visited = new Set();
  let minH = new MinPriorityQueue({priority: (elem) => elem[0]});
  minH.enqueue([0, 0]);
  
  while ( visited.size < N) {
    let [cost, currNode] = minH.dequeue().element;
    if (visited.has(currNode)) {
      continue;
    }
    visited.add(currNode);
    res += cost;
    for (let nextNode = 0; nextNode < N; nextNode++) {
      if (!(visited.has(nextNode))) {
        let nextCost = getWeight(points[currNode], points[nextNode]);
        minH.enqueue([nextCost, nextNode]);
      }
    }
  }
  return res; 
};

function getWeight([xi, yi], [xj, yj]) {
    return Math.abs(xi - xj) + Math.abs(yi - yj); 
}

File: 1584-min-cost-to-connect-all-points/1584-min-cost-to-connect-all-points.py
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        N = len(points)
        adj = {i: [] for i in range(N)}
        
        for i in range(N):
            x1, y1 = points[i]
            for j in range(i+1, N):
                x2, y2 = points[j]
                dist = abs(x1 - x2) + abs(y1 - y2)
                adj[i].append((j, dist))
                adj[j].append((i, dist))
        
        visited = set()
        res = 0
        min_h = [(0, 0)]
        
        while len(visited) < N:
            w, n = heapq.heappop(min_h)
            if n in visited:
                continue
            
            visited.add(n)
            res += w
            
            for n1, w1 in adj[n]:
                if n1 not in visited:
                    heapq.heappush(min_h, (w1, n1))
        
        return res
            
                
        
        